channel/script.go

```diff
@@ -508,48 +674,206 @@ func (ctrl *npcChatController) SendGuildEmblemEditor() {
+	if ctrl.stateTracker.performInterrupt() {
+		ctrl.conn.Send(packetNpcChatUserNumber(ctrl.npcID, text, int32(def), int32(min), int32(max)))
+		ctrl.vm.Interrupt("SendNumber")
+		return def
+	}
+	if len(ctrl.stateTracker.numbers) > ctrl.stateTracker.number {
+		val := ctrl.stateTracker.numbers[ctrl.stateTracker.number]
+		ctrl.stateTracker.number++
+		return int(val)
+	}
+	return def
+}
+
+func (ctrl *npcChatController) SendBoxText(askMsg, defaultAnswer string, column, line int) string {
+	max := column * line
+	if max <= 0 {
+		max = 200
+	}
+	if ctrl.stateTracker.performInterrupt() {
+		ctrl.conn.Send(packetNpcChatUserString(ctrl.npcID, askMsg, defaultAnswer, 0, int16(max)))
+		ctrl.vm.Interrupt("SendBoxText")
+		return defaultAnswer
+	}
+	if len(ctrl.stateTracker.inputs) > ctrl.stateTracker.input {
+		val := ctrl.stateTracker.inputs[ctrl.stateTracker.input]
+		ctrl.stateTracker.input++
+		return val
+	}
+	return defaultAnswer
+}
+
+func (ctrl *npcChatController) SendQuiz(text, problem, hint string, inputMin, inputMax, _ int) string {
+	prompt := text
+	if problem != "" {
+		if len(prompt) > 0 {
+			prompt += "\n"
+		}
+		prompt += problem
+	}
+	if hint != "" {
+		prompt += "\n(" + hint + ")"
+	}
+	if ctrl.stateTracker.performInterrupt() {
+		ctrl.conn.Send(packetNpcChatUserString(ctrl.npcID, prompt, "", int16(inputMin), int16(inputMax)))
+		ctrl.vm.Interrupt("SendQuiz")
+		return ""
+	}
+	if len(ctrl.stateTracker.inputs) > ctrl.stateTracker.input {
+		val := ctrl.stateTracker.inputs[ctrl.stateTracker.input]
+		ctrl.stateTracker.input++
+		return val
+	}
+	return ""
+}
+
+func (ctrl *npcChatController) SendSlideMenu(text string) int {
+	if ctrl.stateTracker.performInterrupt() {
+		ctrl.conn.Send(packetNpcChatSelection(ctrl.npcID, text))
+		ctrl.vm.Interrupt("SendSlideMenu")
+		return -1
+	}
+	if len(ctrl.stateTracker.selections) > ctrl.stateTracker.selection {
+		val := ctrl.stateTracker.selections[ctrl.stateTracker.selection]
+		ctrl.stateTracker.selection++
+		return int(val)
+	}
+	return -1
+}
+
+func (ctrl *npcChatPlayerController) InventoryExchange(itemSource int32, srcCount int32, itemExchangeFor int32, count int16) bool {
+	if !ctrl.plr.removeItemsByID(itemSource, srcCount) {
+		return false
+	}
+
+	item, err := CreateItemFromID(itemExchangeFor, count)
+	if err != nil {
+		return false
+	}
+	if err = ctrl.plr.GiveItem(item); err != nil {
+		return false
+	}
+	return true
 }
 
 func (ctrl *npcChatController) clearUserInput() {
-	ctrl.stateTracker.input = 0
+	// Reset counters but preserve the data
 	ctrl.stateTracker.selection = 0
 	ctrl.stateTracker.input = 0
 	ctrl.stateTracker.number = 0
 }
 
 // Selection value
 func (ctrl *npcChatController) Selection() int32 {
-	val := ctrl.stateTracker.selections[ctrl.stateTracker.selection]
-	ctrl.stateTracker.selection++
-	return val
+	if len(ctrl.stateTracker.selections) == 0 {
+		return -1
+	}
+
+	if len(ctrl.stateTracker.selections) == 1 {
+		ctrl.selectionCalls++
```

Comment:
Can you walk through an example of why this variable is needed? Unless I am missing something you are always returning the last item in the selections slice? Previously it was:
```
func (ctrl *npcChatController) Selection() int32 {
    val := ctrl.stateTracker.selections[ctrl.stateTracker.selection]
    ctrl.stateTracker.selection++ // tracks which selection result we want next
    return val
}
```
which would return the relevant selection for where you were in the script

Looking at this line https://github.com/ErwinsExpertise/Valhalla/blob/master/scripts/npc/1012103.js#L29 I am not following where you mention that the user needs to make 2 selection in the UI as it appears `npc.selection` is called once per avatar ui window.